go/vt/vttablet/onlineddl/executor.go

```diff
@@ -888,6 +891,94 @@ func (e *Executor) initMigrationSQLMode(ctx context.Context, onlineDDL *schema.O
+	constraintMap = map[string]string{}
+	hashExists := map[string]bool{}
+	// newConstraintName generates a new, unique name for a constraint. Our problem is that a MySQL
+	// constraint's name is unique in the schema (!). And so as we duplicate the original table, we must
+	// create completely new names for all constraints.
+	// Moreover, we really want this name to be consistent across all shards. We therefore use a deterministic
+	// UUIDv5 (SHA) function over the migration UUID, table name, and constraint's _contents_.
+	// We _also_ try to include the original constraint name as suffix, if there's enough room.
+	// for example, if the original constraint name is "check_1",
+	// we might generate "chk_7c7de6cb4f9b5842b4d0271f40756883_check_1".
+	// If we then again migrate a table whose constraint name is "chk_7c7de6cb4f9b5842b4d0271f40756883_check_1" we
+	// get for example "chk_f4a9c22556ab5eb788181c0a1e44f248_check_1" (hash changes, and we still try to preserve original name)
+	newConstraintName := func(prefix string, seed string, oldName string) string {
+		if submatch := constraintOriginalNameRegexp.FindStringSubmatch(oldName); len(submatch) > 0 {
+			oldName = submatch[3]
+		}
+
+		hash := textutil.UUIDv5(onlineDDL.UUID, onlineDDL.Table, seed)
+		for i := 1; hashExists[hash]; i++ {
+			hash = textutil.UUIDv5(onlineDDL.UUID, onlineDDL.Table, seed, fmt.Sprintf("%d", i))
+		}
+		hashExists[hash] = true
+		newName := prefix + "_" + strings.Replace(hash, "-", "", -1) + "_" + oldName
+		if len(newName) > maxConstraintNameLength {
+			newName = newName[0:maxConstraintNameLength]
+		}
+		return newName
+	}
+	validateWalk := func(node sqlparser.SQLNode) (kontinue bool, err error) {
+		switch node := node.(type) {
+		case *sqlparser.ForeignKeyDefinition:
+			return false, schema.ErrForeignKeyFound
+		case *sqlparser.ConstraintDefinition:
+			oldName := node.Name.String()
+			var newName string
+			if _, ok := node.Details.(*sqlparser.CheckConstraintDefinition); ok {
+				newName = newConstraintName("chk", sqlparser.CanonicalString(node.Details), oldName)
+			} else if _, ok := node.Details.(*sqlparser.ForeignKeyDefinition); ok {
+				newName = newConstraintName("fk", sqlparser.CanonicalString(node.Details), oldName)
+			}
+			if newName != "" {
+				node.Name = sqlparser.NewColIdent(newName)
+				constraintMap[oldName] = newName
+			}
+		}
+		return true, nil
+	}
+	if err := sqlparser.Walk(validateWalk, createTable); err != nil {
+		return constraintMap, err
+	}
+	return constraintMap, nil
+}
+
+// validateAndEditAlterTableStatement inspects the AlterTable statement and:
+// - modifies any CONSTRAINT name according to given name mapping
+func (e *Executor) validateAndEditAlterTableStatement(ctx context.Context, alterTable *sqlparser.AlterTable, constraintMap map[string]string) (err error) {
+	validateWalk := func(node sqlparser.SQLNode) (kontinue bool, err error) {
+		switch node := node.(type) {
+		case *sqlparser.DropKey:
```

Comment:
Side note: I would have expected this to be DropConstraint — for FK and CHECK constraints — as they are rules rather than indexes, but I see we don't have one. Perhaps it's not needed...